### ISSUE TITLE
stream.hls: stop stream early when EXT-X-ENDLIST is missing

### ISSUE THREAD

**Issue by earthtone2** [streamlink/streamlink#2198]
When a live stream ends, streamlink retries 12 times and then exits. (12 is a strange number and I'm not sure where it comes from. `hls-segment-attempts` is default 3, so maybe the code that uses it is being done 4 times, for a total of 12. That's just stabbing in the dark though.)

I would like to limit that to only 1 retry, because I'm worried about hitting the server too much. I was once IP banned from a server for using an automated download program, and it now makes me worry about making too many attempts-- especially attempts which are clearly automated.

Ideally, when a live stream ends, I'd like to wait 60 seconds, then retry just once. Is there a way to do that?

The message I get when a stream ends is `[stream.hls][warning] Failed to reload playlist: Unable to open URL: ... (403 Client Error: Forbidden for url:..)`, which comes from hls.py.

This is streamlink 0.14.2 on both boxes I used (Ubuntu and Windows).

---

Update: I just tried this on my Windows box, testing by turning the internet off. The first time it made 8 attempts; the second time it made 12. This was while using `--hls-playlist-reload-attempts=1`.

Update2: I actually timed the time it took to finally die and it was 38 seconds. Combined with the dead airtime before the error messages start, that's getting close to 60 seconds. Is `hls-timeout` what I need?

Update3: It looks like the time it takes can be controlled with `hls-timeout`, and these 8-12 attempts I am seeing were actually just a loop of attempts until that time runs out. So it looks like I can prevent _any_ attempts by setting that to a low value, like 15.0. That's better than nothing, although I'd still rather have it pause for a minute, then exit if it gets another 403.

**Comment by bastimeyer** [comment]
Since this was just mentioned in another thread and since I've refactored the HLS code a bit recently, what you've said regarding `--hls-timeout` (removed in favor of the generic `--stream-timeout` parameter) is correct. This parameter sets the read timeout for the output buffer in [`SegmentedStreamReader`](https://github.com/streamlink/streamlink/blob/db73cb2b00efaf54a13d6a6c642f49ea4777c6b2/src/streamlink/stream/segmented.py#L207-L236), which is 60 seconds by default. If no data was written to the buffer and there was a read attempt by the stream consumer, [this read attempt will time out](https://github.com/streamlink/streamlink/blob/master/src/streamlink/buffers.py#L104-L109) after these 60 seconds and an `OSError` will be raised with a "read timeout".

Another way of making Streamlink stop after a certain period of time when a stream has stopped but no [`#EXT-X-ENDLIST`](https://datatracker.ietf.org/doc/html/rfc8216#section-4.3.3.4) tag was set could be checking for new HLS segments to be queued. The HLSStreamWorker will continue refreshing the playlist when this tag is missing, but if there are no new segments (sequence number incrementation), this could be checked and the stream be ended this way when a certain time threshold has been reached. This would not be a graceful way of ending the stream, but the read timeout isn't either, and it's the responsibility of the server to include the endlist tag to signal the actual end of the stream.

**Comment by bastimeyer** [comment]
I have a local branch here which will make Streamlink end the stream early when the HLS playlist didn't contain new segments for an X amount of time. This threshold is currently set to `TARGETDURATION * 2`.

That factor `2` is chosen very liberally, despite the HLS spec saying that the playlist's [`TARGETDURATION`](https://datatracker.ietf.org/doc/html/rfc8216#section-4.3.3.1) is the max duration of any segment, which means that when this time has passed, new segments MUST be available. The factor could maybe be reduced to 1.5 or so, but what I want to avoid are potential issues during stream discontinuities like embedded ads, where the targetduration could change to a different value that could result in the stream ending early by accident. The segment-queue timing threshold is independent of the playlist reload time since the reload time can be changed via session options and via custom stream implementations, like Twitch low latency streaming. I'm also not sure yet if I want to add a new session option that can change the value.

Since the changes depend on #5321 which is currently unmerged, I can't open a new PR for that just yet.
